### ISSUE TITLE
encryption: changing key provider name breaks it

### ISSUE THREAD

**Issue by aviau** [opentofu/opentofu#1605]
### OpenTofu Version

```shell
OpenTofu v1.7.0
```

### OpenTofu Configuration Files

Changing from

```hcl
    key_provider "aws_kms" "aes_256" {
      kms_key_id = "mrk-1"
      region     = "us-east-1"
      key_spec   = "AES_256"
    }
```

to

Changing from

```hcl
    key_provider "aws_kms" "hello" {
      kms_key_id = "mrk-1"
      region     = "us-east-1"
      key_spec   = "AES_256"
    }
```

Breaks the config and will result in `Error: error loading state: decryption failed for all provided methods: attempted decryption failed for state: no decryption key available`

### Debug Output

.

### Expected Behavior

This is unexpected for terraform config, since when does resource naming change anything?

### Actual Behavior

.

### Steps to Reproduce

tofu apply

**Comment by aviau** [comment]
> key providers need to be able to store metadata alongside the encrypted data

This part makes sense.

> the only way we can uniquely identify a key provider is by its name

I lost you at this second part of your sentence.

The state can only be encrypted with a single provider at a time, right? So why would the name of the provider matter? You need one set of metadata associated with the state at any given time.

Quick story:
- Yesterday I was migrating a dozen states to encryption, I have code that generates my `terraform.tf`, after a while I had ran so many `terraform apply` that I ended up forgetting which state was encrypted with which provider name. I found it quite ridiculous that recovering my state required remembering what the name of the key provider was.

**Comment by aviau** [comment]
(dont worry for my states, I could recover them, but I still think there is user experience improvement potential here).

Do think that a more tofu-y solution to this problem would be to add an `id` configuration to providers?

OpenTofu users are not used to the fact that resource names matter, but they are used to creating cloud resources that have unique names. However these unique names are usually fields of the resource.

**Comment by aviau** [comment]
I agree that defaulting to the resource name is a compromise, but at the same time there is no other resource that does that.

This makes this feature different than everything else in OpenTofu, I think making it required would be better because it makes it more discoverable / self-explained. I can foresee this being a source of confusion forever.

At the same time I understand you probably want to avoid merging breaking changes at this time.

**Comment by aviau** [comment]
Also, does it mean the provider name needs to be the same when you read from a remote state?

**Comment by aviau** [comment]
That’s unfortunate. When you work on an OpenTofu environment ideally you don’t have to think about avoiding clashes with names from future environments that will read from it, it prevents modular thinking / encapsulation.

**Comment by StephanHCB** [comment]
I am a bit torn on this.

On the one hand it is perfectly normal tofu behaviour to force recreation if the name of a resource changes. Or even if the index in a for_each changes.

On the other hand, the cases where multiple key providers exist to accomodate multiple remote state data sources are rare, and it might be ok to just try the available configuration for decryption of a given state type, maybe with a configuration field to enable this implicit fallback. The encrypted states have a mac, so we can detect when the decryption result is invalid, so there should be little harm done other than a bit of extra time spent attempting decryption. And a fair bit of ease of use benefit.

**Comment by cube2222** [comment]
> I agree that defaulting to the resource name is a compromise, but at the same time there is no other resource that does that.

Isn't that how most resources in Tofu work? If you change a resource's name, you'll get a destroy+create in the next plan, unless you use a `move` block. The corresponding procedure here is going through a two-step process with a fallback block.

> When you work on an OpenTofu environment ideally you don’t have to think about avoiding clashes with names from future environments that will read from it.

Agree with this.

> that's a good point, one that could be remedied by using the suggested IDs.

Would it? That is, you'd still have to think about making the IDs unique, no? Or is the point that two key providers with different names, but the same ID (let's say `key`), will not locally conflict (different names) but still be possible to be used with two separate remote statefiles that each expect a `key` key_provider?

Overall, I agree we should think hard about how we want this UX to look like, and make it as simple and non-surprising as possible.

Presumably, if you use a key_provider called `key` in all your configs, and it always references the same KMS key, then you could use this single key_provider for both your local state, and all remote state reads, right? For local decryption it would use metadata from the local state-file, while for remote decryption it would get the metadata from the remote state-file?

**Comment by cube2222** [comment]
Right now we only allow configuring everything related to state-encryption globally, right?

If we allowed an encryption block inside of a terraform_remote_state datasource, that could contain all the usual (key provider, method) and would be completely isolated from the "main" encryption config. Would that also possibly solve the problem of conflicts?

**Comment by aviau** [comment]
> Isn't that how most resources in Tofu work?

In Tofu, the name of the resource cannot be used by the resource itself.

When I create an AWS bucket, I can name it whatever I want, and this information is not exposed to the resource itself.

When I am programming, I should be allowed to name my variables however I want, and it shouldn't affect anything.

In this config, you are using the variable name (resource name) as a hidden parameter to the resource.

**Comment by aviau** [comment]
> If we allowed an encryption block inside of a terraform_remote_state datasource

Thats the most simple idea IMO, and It should have been like that from the start. Why is the config split in half anyways? When I saw the docs I assumes it was because of a Tofu/Opentofu cross-compatibility challenge. The design of having an encryption block that points to the terraform_remote_state address was not the most intuitive.

**Comment by cube2222** [comment]
> In Tofu, the name of the resource usually cannot be used by the implementation of the resource. It decides where the resource is stored in the state, but it isn't an argument of the resource.

This is actually exactly how it's used here as well - how it's stored in the state-file. Each key-provider has its own entry in the state-file, keyed by its name. But I do get your point. We'll definitely want to address this.

**Comment by cam72cam** [comment]
I've created https://github.com/opentofu/opentofu/issues/1641 to track the ability to supply encryption configuration directly in data blocks.  A decision needs to be made there on functionality, please take a look and share your thoughts!

**Comment by cube2222** [comment]
The core team (on an eng sync) agreed that it makes sense to add an `id` field to key providers. Only thing now is how it should be named :D Just `id` is too general, and doesn't indicate that this is about how the metadata is keyed in the statefile/planfile.

I would suggest `metadata_storage_key`, which I think carries the meaning very explicitly.

**Comment by brettcurtis** [comment]
It's still unclear if I should be able to do what @Evi1Pumpkin had in: https://github.com/opentofu/opentofu/issues/1978 with a 1.9+ release?

**Comment by yottta** [comment]
If you refer to reading states by using `terraform_remote_state` data block, then yes, it should work fine. This is documented in the [state encryption page](https://opentofu.org/docs/language/state/encryption/#remote-state-data-sources).

EDIT: Just to sure that this is clear, I added a comment in [#1978](https://github.com/opentofu/opentofu/issues/1978#issuecomment-3569423178). Please let us know in that issue's comments section if you need more help.
